src/neuralnet/beacon.cpp

```diff
@@ -185,34 +282,52 @@ void BeaconRegistry::Add(Contract contract)
 
     payload.m_beacon.m_timestamp = contract.m_tx_timestamp;
 
-    m_beacons[payload.m_cpid] = std::move(payload.m_beacon);
+    // Legacy beacon contracts before block version 11--just load the beacon:
+    //
+    if (contract.m_version == 1) {
+        m_beacons[payload.m_cpid] = std::move(payload.m_beacon);
+        return;
+    }
+
+    // For beacon renewals, check that the new beacon contains the same public
+    // key. If it matches, we don't need to verify it again:
+    //
+    if (TryRenewal(m_beacons, payload)) {
+        return;
+    }
+
+    // Otherwise, set the new beacon aside for scraper verification. The next
+    // superblock will activate it if it matches a BOINC account:
+    //
+    PendingBeacon pending(payload.m_cpid, std::move(payload.m_beacon));
+
+    m_pending.emplace(pending.GetId(), std::move(pending));
 }
 
 void BeaconRegistry::Delete(const Contract& contract)
 {
     const auto payload = contract.SharePayloadAs<BeaconPayload>();
 
-    m_beacons.erase(payload->m_cpid);
-}
+    if (contract.m_version >= 2) {
+        m_pending.erase(payload->m_beacon.GetId());
+    }
 
-void BeaconRegistry::Revert(const Contract& contract)
-{
-    IContractHandler::Revert(contract);
+    m_beacons.erase(payload->m_cpid);
 }
 
 bool BeaconRegistry::Validate(const Contract& contract) const
 {
-    // Only administrative contracts can delete beacons. This is verified
-    // by master public key when checking the contract.
-    //
-    if (contract.m_action == ContractAction::REMOVE) {
-        return true;
-    }
-
     // For legacy beacons, check that the unused parts contain non-empty values
     // for compatibility with the existing protocol to prevent a fork.
     //
     if (contract.m_version <= 1) {
+        // Only administrative contracts can delete beacons. This is verified
```

Comment:
I think this comment is now inaccurate

Reply:
It still applies to legacy contracts until the Fern protocol kicks-in.

Reply:
maybe `Only administrative contracts can delete legacy beacons` or something like that instead?

Reply:
Done.